### ISSUE TITLE
Interesting, but tooth or noise?

### ISSUE THREAD

**Issue by gillonba** [KazuhideMimura/yolov5-ichthyolith#5]
Interesting project, I have recently been experimenting with something similar because building a yolo dataset for object detection is hard, but classification is easy.  As a bonus, down the line it should be easy to write a script to build a dataset to train a single stage yolo model based on detection+classification of any number of source images.  Question though, why is the tooth-transverse in the second detection identified as noise in the other two detection runs?

**Comment by KazuhideMimura** [comment]
Hi, thank you for important question that were not written in readme.
The score for a rough class is calculated by summing the scores of the subclasses that belong to the rough class.

Therefore, if the classification model predicted as follwos, the top-1 fine class will be Tooth-A, but top-1 coarse class will be Noise.
> Tooth-A: 40 %
> Noise-B: 30%
> Noise-C: 30%

This is implemented in lines 926-935 of utils/general.py as follows:

    # coarse-class determination
    if reclass_dict is None:
        # keep detailed classes
        PRED_cls3 = PRED_cls2
    else:
        # determine coarse class and confidence by sum of detailed-classes scores
        PRED_cls3 = torch.zeros([PRED_cls2.size()[0], len(reclass_dict)], device=device_)
        for k, v in reclass_dict.items():
            PRED_cls3[:, k] = PRED_cls2[:, v].sum(1)            
    pred_conf3, pred_cls3 = PRED_cls3.max(1)

**Comment by KazuhideMimura** [comment]
I found that several important files were missing in the utils folder, which are now uploaded.
Thank you for letting me know this.

**Comment by gillonba** [comment]
Has this approach been working for you?  It seems like simply summing the subclass scores has some issues, most obviously if one class has more subclasses, or by accidentally hitting a high score for an image that is outside the scope of the original training, simply because it ended up in the wrong class.  

In my approach I am using yolo5 to do detection, but then clipping the detections and using Microsoft's ML.Net image classifier to classify the clipped images.  It's relatively slow but it's flexible, allows me to chain classifications, and makes it super easy to build classification datasets.  Also, I'm a .Net developer by day so doing it all through a python script isn't my go-to

**Comment by KazuhideMimura** [comment]
>Has this approach been working for you?

Determining rough classes is actually not working effectively in my project. It may be because the classification model is not fully trained. I think I need to prepare more training dataset for classification.

> if one class has more subclasses, or by accidentally hitting a high score for an image that is outside the scope of the original training

This is very interesting point of view that I need to consider when I get well-trained classification model. Thank you for the advice.

Your approach also seems interesting. I have also tried combining two libraries (Mask RCNN + EfficientNet-V2). But in my project, since I need to detect from ~10,000 microscopic images (or more!), detection time is very critical for efficiency.
https://github.com/KazuhideMimura/ai_ichthyolith, https://github.com/KazuhideMimura/eNetV2_for_ai_ichthyolith

Thank you for the comment anyway!
